Explain focus problems and description of algorithms for LCT module

Gabriel-Xu opened the pull request:
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

SansPapyrus683 commented:
this isn't meant to be complete right

Gabriel-Xu commented:
> this isn't meant to be complete right

I'm not sure exactly what you are looking for since the instructions said to explain the focus problems and describe the algorithms. There was one focus problem that I thought didn't require much explanation if you understand the material provided in the resources. I have seen many currently complete modules leave the explanation of algorithms mostly to the attached resources, so I could explain the part about sum/max/min over paths more, but otherwise I'm not sure what part I should add to.

ryanchou-dev commented:
Could you explain the sample problems on the module? (like LCT - Subtrees)

Let me know if you have questions!

Gabriel-Xu commented:
> Could you explain the sample problems on the module? (like LCT - Subtrees)
> 
> Let me know if you have questions!

Does explaining the problem mean adding a code snippet in addition to a text explanation? Also, does sample problems refer specifically to the Vertex Add Path Sum problem and Vertex Add Subtree Sum problem?

ryanchou-dev commented:
Yep.

ryanchou-dev commented:
any updates?

Gabriel-Xu commented:
> any updates?

I am working on it

Gabriel-Xu commented:
Could someone review my new commits?

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -399,6 +657,224 @@ int main() {
+typedef long long ll;
+
+// BeginCodeSnip{Benq Link Cut Tree}
+/**
```
if this impl is linked in the resources, remove this doc comment

Gabriel-Xu replied:
Fixed

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -399,6 +657,224 @@ int main() {
 
+### Implementation
+
+We can also maintain information about subtrees by keeping track of values for the virtual subtrees of a node. When querying for information such as the subtree sum, we call $\textup{access}()$ on the node so that all of its children in the represented tree are part of virtual subtrees and then retrieve the desired value.
```
uh i think we use texttt instead of textup

Gabriel-Xu replied:
Fixed

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -357,14 +388,239 @@ int main() {
+};
+// EndCodeSnip
+
+const int maxn = 2e5 + 5;
```
rename to MAX_N
also is the +5 necessary

Gabriel-Xu replied:
Fixed, the +5 is not necessary

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -357,14 +388,239 @@ int main() {
+typedef long long ll;
+
+// BeginCodeSnip{Benq Link Cut Tree}
+/**
```
same thing w the comment here

Gabriel-Xu replied:
Fixed

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -399,6 +637,204 @@ int main() {
+		cin >> u >> v;
+		link(LCT[u], LCT[v], 1);
+	}
+	while (q--) {
```
can you use a for loop instead (and the other sol too)

Gabriel-Xu replied:
Fixed

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -41,6 +44,20 @@ frequency: 1
+  to make a new tree
+- Find the root of the tree that a node belongs to
+
+These operations all use the $\text{access}(v)$ subroutine, which creates a preferred path from the root of the represented tree to vertex $v$, making a corresponding auxiliary splay tree with $v$ as the root.
```
```suggestion
These operations all use the $\texttt{access}(v)$ subroutine,
which creates a preferred path from the root of the represented tree to vertex $v$,
making a corresponding auxiliary splay tree with $v$ as the root.
```

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -357,14 +388,219 @@ int main() {
+			LCT[p]->access();
+			LCT[p]->val += x;
+			LCT[p]->calc();
+		} else {
```
```suggestion
		} else if (t == 2) {
```

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -399,6 +637,204 @@ int main() {
+			LCT[p]->access();
+			LCT[p]->val += x;
+			LCT[p]->calc();
+		} else {
```
```suggestion
		} else if (t == 2) {
```

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -386,6 +624,8 @@ int main() {
 
+### Sample Problem
+
 <Problems problems="subSam" />
```
make this a focus problem?

Gabriel-Xu replied:
Done

SansPapyrus683 commented:
i made some changes myself, please be sure to pull!

Gabriel-Xu commented:
> i made some changes myself, please be sure to pull!

Okay, your changes are all added.

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -355,16 +388,226 @@ int main() {
+
+The link cut tree can be used to solve problems that deal with queries updating trees and querying for connectivity.
+
+### Problems
 
 <Problems problems="con" />
```
@ryanchou-dev this part is just two problems...i'm thinking smth like either delete both or keep just one and explain it

ryanchou-dev replied:
@Gabriel-Xu are you cool with explaining one of these problems briefly?

i think explanation would be better than just deleting the section

Gabriel-Xu replied:
> @Gabriel-Xu are you cool with explaining one of these problems briefly?
> 
> i think explanation would be better than just deleting the section

Do you want it to be another focus problem?

Gabriel-Xu replied:
Also does this mean I should include a code solution or just a text explanation?

Gabriel-Xu replied:
For now, I added a brief explanation.

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
```diff
@@ -215,7 +233,22 @@ int main() {
-### With Euler-Tour Tree
+### With Euler Tour Tree
+
+An alternative solution is to use the Euler Tour Tree (ETT) technique.
```
i believe the usaco guide has a module on the euler tour, can you link that?

Gabriel-Xu replied:
Should I add it as a resource or just a normal hyperlink?

SansPapyrus683 replied:
just a normal hyperlink
but i'm dumb so can you check if this euler tour is the same as the one covered in the guide? if it isn't then add an explanation on how this differs from the other euler tour

Gabriel-Xu replied:
Done

SansPapyrus683 commented on content/6_Advanced/LCT.mdx:
````diff
@@ -355,16 +388,226 @@ int main() {
 
+**Time Complexity:** $\mathcal{O}(?)$
+
+```cpp
````
around all cpp code, please wrap it with `<LanguageSection>` and `<CPPSection>`
check the other modules for how they do stuff

Gabriel-Xu replied:
Done

SansPapyrus683 commented:
alright can you test the updated solutions and show if they still work

Gabriel-Xu commented:
> alright can you test the updated solutions and show if they still work

Yes, the solutions work

SansPapyrus683 commented:
may i see proof?

Gabriel-Xu commented:
> may i see proof?

https://judge.yosupo.jp/submission/206454
https://judge.yosupo.jp/submission/206455
![AC](https://github.com/cpinitiative/usaco-guide/assets/105399165/4fcada64-a9b3-432c-8697-df45b1e2e741)

SansPapyrus683 reviewed (approved):
i'm allowing the benq ahh data structure only because i haven't made a sansified version yet

ryanchou-dev reviewed (commented):
otherwise all good!

Gabriel-Xu commented:
Can someone please review this?

ryanchou-dev commented:
> Can someone please review this?

Looks good. I think the intention was for the problem to be a focus problem with a brief explanation below. If you already have code, you could add that as well.

You can delete the "Silence's Sounds" problem. Everything else looks great, so I'll merge after that 🙏

Gabriel-Xu commented:
> > Can someone please review this?
> 
> Looks good. I think the intention was for the problem to be a focus problem with a brief explanation below. If you already have code, you could add that as well.
> 
> You can delete the "Silence's Sounds" problem. Everything else looks great, so I'll merge after that 🙏

Okay, I made it a focus problem and added the code, and deleted the "Silence's Sounds" problem.

Gabriel-Xu commented:
@ryanchou-dev  is it good?

Gabriel-Xu commented:
Please take a look soon

ryanchou-dev reviewed (approved):
perfect! thanks for all of your contributions :)

Gabriel-Xu commented:
Thanks!

SansPapyrus683 commented:
uh so we can pay through paypal or amazon gift card
send email w/ the relevant info to [handle] AT gmail.com
[handle] = kevinsheng2005
